evap/evaluation/tools.py

```diff
@@ -271,30 +276,52 @@ def is_external_email(email):
     return not any([email.endswith("@" + domain) for domain in settings.INSTITUTION_EMAIL_DOMAINS])
 
 
-def user_publish_notifications(courses):
-    user_notifications = defaultdict(set)
-    for course in courses:
+def user_publish_notifications(grade_document_courses, evaluation_results_courses):
+    publish_notifications = defaultdict(lambda: CourseLists(set(), set()))
+
+    for course in evaluation_results_courses:
         # for published courses all contributors and participants get a notification
         if course.can_publish_grades:
             for participant in course.participants.all():
-                user_notifications[participant].add(course)
+                publish_notifications[participant].evaluation_results_courses.add(course)
             for contribution in course.contributions.all():
                 if contribution.contributor:
-                    user_notifications[contribution.contributor].add(course)
+                    publish_notifications[contribution.contributor].evaluation_results_courses.add(course)
         # if a course was not published notifications are only sent for contributors who can see comments
         elif len(course.textanswer_set) > 0:
             for textanswer in course.textanswer_set:
                 if textanswer.contribution.contributor:
-                    user_notifications[textanswer.contribution.contributor].add(course)
-            user_notifications[course.responsible_contributor].add(course)
+                    publish_notifications[textanswer.contribution.contributor].evaluation_results_courses.add(course)
+            publish_notifications[course.responsible_contributor].evaluation_results_courses.add(course)
+    for course in grade_document_courses:
+        # all participants who can download grades get a notification
+        for participant in course.participants.all():
+            if participant.can_download_grades:
+                publish_notifications[participant].grade_document_courses.add(course)
+
+    return publish_notifications
+
+
+def send_publish_notifications(grade_document_courses=[], evaluation_results_courses=[]):
+    publish_notifications = user_publish_notifications(grade_document_courses, evaluation_results_courses)
+
+    for user, course_lists in publish_notifications.items():
+        try:
+            EmailTemplate.get_publish_template().send_to_user(
+                user, 
+                grade_document_courses=list(course_lists.grade_document_courses),
+                evaluation_results_courses=list(course_lists.evaluation_results_courses)
+            )
+        except Exception:
+            logging.getLogger(__name__).exception(("An error occured when sending the notification email to {}.").format(user.username))
```

Comment:
i suggest putting this around the actual send_mail, which is the place where we kinda "expect" exceptions.